[8기 주재성] TodoList with Ajax

JaeSeongJu opened the pull request:
Ajax에 대해서 공부하고, 전 과제에서 연결시키는 방법을 고민하다보니 기능 구현까지만 급하게 완료했습니다. 
리팩토링은 물론 전체적으로 코드가 정리가 전혀 안된 것 같습니다...ㅠㅠ

엉성하지만 다른 분들의 코드 많이 참고해서 무엇을 공부해야 할지 잘 배워가도록 하겠습니다!

## 🎯 요구사항

-   [x] 1. User 추가하기
-   [x] 2. User의 투두리스트 불러오기
-   [x] 3. User 삭제하기
-   [x] 3. todoItem 추가하기
-   [x] 4. todoItem 불러오기
-   [x] 5. todoItem complete하기
-   [x] 6. todoItem 삭제하기
-   [x] 7. todoItem contents 내용 수정하기

<br/>

## 🎯🎯 심화 요구사항

-   [x] 1. fetch api 사용하는 부분을 async await을 사용하여 리팩토링하기.
-   [x] 2. github issue에서 라벨을 붙이는 것처럼, 우선순위에 따라서 label를 추가하기.
-   [x] 3. ES6 impot & export를 이용해 자바스크립트 파일을 리팩토링하기.

<br/>

## 🕵️‍♂️ 제약사항

-   [x] 1. User의 이름은 최소 2글자 이상이어야 한다.
-   [x] 2. TodoItem Contents는 최소 2글자 이상이어야 한다.

JaeSeongJu committed:
feat : 유저 추가
- 유저 생성 버튼 클릭 후 이름 입력시 POST로 유저 추가 후 유저 버튼 랜더링
- 유저 정보 담은 리스트 업데이트

JaeSeongJu committed:
feat & fix : User 목록 불러오기
- 유저 목록 불러온 뒤 유저 추가 기능까지 되도록 기능 구현

JaeSeongJu committed:
feat & fix : 유저별 todoList 불러오기 기능 추가 / todoList 기능 추가
- 이전 미션 기능 불러온 뒤 object key 값 수정
- 각 유저별 저장된 todoList 정보를 불러와서 랜더링

JaeSeongJu committed:
feat & refactor : todoItem 공통 저장 리스트 생성, user 삭제 기능
- 각 유저별 todoItem을 공통 list에 저장해서 기능 동작하게 만들었음

JaeSeongJu committed:
feat : complete 여부 설정

JaeSeongJu committed:
feat : todoItem contents 수정 & 유저별 todoList title 붙이기

JaeSeongJu committed:
feat : loading창 띄우기

JaeSeongJu committed:
feat : todo Item 전체 삭제 기능 추가

JaeSeongJu committed:
feat : todoItem 우선순위 설정 기능 추가
- Object 사용으로 switch나 if문 조건이 많아지는 것을 대체

bbbnam commented on src/js/API.js:
```diff
@@ -0,0 +1 @@
+export const BASEURL = "https://js-todo-list-9ca3a.df.r.appspot.com";
```
js 파일명은 API 라서 뭔가 api 호출 등을 할거 같은 파일인데
상수만 하나 있군요..😅
constant.js 와 같은 네이밍을 사용하거나, 해당 API url을 호출하는 곳 상단에 선언해주어도 충분하지 않을까요?

JaeSeongJu replied:
현 파일에 API url들을 한 번에 모아서 관리해 줘야겠어요!!!

bbbnam commented on src/js/Todo.js:
```diff
@@ -0,0 +1,315 @@
+	todoList.addEventListener("dblclick", editItem);
+	todoList.addEventListener("keyup", finishEdit);
+	todoList.addEventListener("change", selectPriority);
+}
```
이벤트 리스너들을 한곳에서 관리하니 깔끔하네요 😁👍

bbbnam commented on src/js/Todo.js:
```diff
@@ -0,0 +1,315 @@
+			const user = document.querySelector(".active");
+			const addedItem = await fetchAddItem(user.dataset.id, inputText);
+			addItem(addedItem._id, inputText, false, "NONE");
+			todoInput.value = "";
```
1주차 공통 피드백 내용 중 '빠른 실패' 를 사용해서 if ~ else문을 개선해 보는건 어떨까요?🤔

JaeSeongJu replied:
'빠른 실패' 생각은 항상 하는데 구현은 실패했네요 ㅠㅠ 노력해보겠습니다 !

JaeSeongJu replied:
'빠른 실패' 꼭 기억하겠습니다 🥲

bbbnam commented on src/js/Todo.js:
```diff
@@ -0,0 +1,315 @@
+	}
+	if (event.target === pendingBtn) {
+		renderTodoItem(pendingList);
+	}
```
if문이 반복되는 부분은 object literal 을 사용해서 변경해보는 건 어떨까요?😁

JaeSeongJu replied:
오! 이 부분도 object literal을 사용하면 더 깔끔하게 만들 수 있겠네요!

bbbnam commented on src/js/Todo.js:
```diff
@@ -0,0 +1,315 @@
+			"Content-Type": "application/json",
+		},
+		body: JSON.stringify({ priority }),
+	});
```
fetch (url , option)의 구조가 반복되고 있는데요, 이 부분을 공통 코드로 뺄 수 있지 않을까요?😃

JaeSeongJu replied:
API url 관리처럼 fetch도 따로 파일을 생성해서 공통 함수를 만들어 구현해주면 좋을 것 같아요!

bbbnam commented on src/js/User.js:
```diff
@@ -0,0 +1,139 @@
+		)
+			userBtn.classList.remove("active");
+		else if (userBtn.dataset.id === clickedBtn.dataset.id)
+			userBtn.classList.add("active");
```
중괄호를 생략하면 가독성이 좋아지기도 하지만,
되도록이면 생략하지 않는게 추후 작업 추가나, 유지보수시에 헷갈리지 않을 수 있을 거 같습니다.
https://helloinyong.tistory.com/15 이걸 한번 참조해보시면 좋을 거 같아요.😁

JaeSeongJu replied:
링크해 주신 내용도 잘 읽어보았습니다!🙇‍♂️
좋은 코드는 물론, 많은 부분에 있어서 조언 감사드립니다 자룡님 ! 🙂

bbbnam reviewed (commented):
저는 이번에 구현도 제대로 못해서
재성님의 코드를 보고 많이 배워갑니다-!

깔끔하게 참 잘 짜신거 같아요😃😍👍👍

jung0930 commented on src/js/Todo.js:
```diff
@@ -0,0 +1,315 @@
+const PENDING = "false";
+const COMPLETED = "completed";
+
+const todoInput = document.getElementById("new-todo-title");
+const todoList = document.getElementById("todo-list");
+const todoCount = document.querySelector(".todo-count strong");
+
+const showAllBtn = document.querySelector(".all");
+const completedBtn = document.querySelector(".completed");
+const pendingBtn = document.querySelector(".active");
+const deleteAllBtn = document.querySelector(".clear-completed");
```
document.querySelector는 반복되서 사용되니 줄여서 쓰는건 어떨까요?
const $ = (selector) => document.querySelector(selector);
const todoCount = $(".todo-count strong");

JaeSeongJu replied:
```javascript
const $ = (selector) => document.querySelector(selector);
```
처음 보는 방법이에요! 깔끔하게 dom 요소들을 불러올 수 있네요! 사용해보겠습니다!!

jung0930 commented on src/js/Todo.js:
```diff
@@ -0,0 +1,315 @@
+			item.priority
+		);
+	});
+	todoList.innerHTML = mergedTemplate.join("");
```
템플릿용 메서드를 만들어두고 반환해서 그려주는 방식 깔끔하게 잘짜신거같습니다. 배우고갑니다!

jung0930 commented on src/js/Todo.js:
```diff
@@ -0,0 +1,315 @@
+	// loadData();
+	// todoItemList = getUserTodoList();
+	todoInput.addEventListener("keydown", enterItem);
+	showAllBtn.addEventListener("click", showProgress);
+	completedBtn.addEventListener("click", showProgress);
+	pendingBtn.addEventListener("click", showProgress);
```
showProgress 메서드에서 상태별 보기 버튼에 따른 분기처리가 이미 되어 있는거 같은데
모든버튼에 한번에 addEventListener를 해주는건 어떨까요?

JaeSeongJu replied:
버튼에 공통 클래스를 만들어 한 번에 이벤트를 등록하는 것을 말하시는 건가요? 
세부적인 내용들이 공통 이벤트 핸들러에 처리가 되어있으니 좋은 것 같아요!!
이벤트를 showProgress로 한 번에 처리하기는 했는데, 모든 버튼에 한 번에 주는 방법은 저렇게만 떠오르네요 ㅠㅠ

jung0930 commented on src/js/User.js:
```diff
@@ -0,0 +1,139 @@
+
+const onUserCreateHandler = async (event) => {
+	const userName = prompt("추가하고 싶은 이름을 입력해주세요.");
+	if (userName && userName.length >= 2) {
```
prompt로 받은 boolean값을 if문에 넣으니 더 깔끔하고 가독성이 좋은거같습니다. 배우고 갑니다 :)

jung0930 commented on src/js/User.js:
```diff
@@ -0,0 +1,139 @@
+		const userData = await addUser(userName);
+		updateUser(userData);
+		const newUser = users[users.length - 1];
+		renderUserBtn(newUser);
+	} else alert("두 글자 이상으로 적어주세요!");
```
else문을 작성하지않고 빠른 실패를 하게하면 코드의 의도와 예외 조건이 명확해지고 나중에 분리하기도 쉽다고 1주차 공통 피드백 내용 중에 있더라구요.
저는 필요없는 else문이라도 적어두는게 맞다/아니다 가독성이 더 좋은거 같은데 어떻게 생각하실까요?

JaeSeongJu replied:
빠른 실패나 에러 핸들링을 통해서 예외 요소들을 효율적으로 잘 잡아내는 것이 중요한 것 같습니다.
저는 아직 미숙한 것 같습니다😂
그래서 else문이 필요 없다는 생각이 안들도록 코드를 구현하는 것이 좋은 것 같다고 생각합니다!

jung0930 reviewed (commented):
좋은 리뷰를 남겨드리지 못한거 같아 죄송합니다ㅠㅠ
주재성님 깔끔한 코드보면서 한수 배우고 갑니다. 감사합니다. :)

JaeSeongJu committed:
refactor : API 관리 파일 분리
- API.js 파일을 만들어, url과 fetch를 User, Todo로 Object형태로 분류해서 구현

JaeSeongJu committed:
refactor : Dom selector, HTML Template 관리 파일 분리
- dom selector를 함수 형식으로 바꿈

JaeSeongJu committed:
fix : 피드백 반영 일부 수정
- '$' 함수 생성해서 dom element 가져오기, $변수 앞에 붙여서 dom 요소 나타내기
- 빠른 실패로 if문 순서 변경
+ UserTodoList 저장하는 부분 data store / extract 용도로 바꾸면 더 깔끔해질 것 같음
